Source/Core/Core/PowerPC/JitArm64/JitArm64_LoadStore.cpp

```diff
@@ -755,6 +755,90 @@ void JitArm64::stmw(UGeckoInstruction inst)
+// While it does automatically decrease the downcount and CTR, it does *not* automatically perform
+// `rX += 32 * WA`.
+//
+// All the passed registers will be clobbered.
+//
+// WA will be set to the number of loops to execute minus 1.
+void JitArm64::WriteInitCacheLoop(const u32 cycle_count_per_loop, const ARM64Reg loop_counter,
+                                  const ARM64Reg WA, const ARM64Reg WB)
+{
+  // We'll execute somewhere between one single cacheline invalidation and however many are needed
+  // to reduce the downcount to zero, never exceeding the amount requested by the game.
+  // To stay consistent with the rest of the code we adjust the involved registers (CTR and Rb)
+  // by the amount of cache lines we invalidate minus one -- since we'll run the regular addi and
+  // bdnz afterwards! So if we invalidate a single cache line, we don't adjust the registers at
+  // all, if we invalidate 2 cachelines we adjust the registers by one step, and so on.
+
+  const auto reg_cycle_count = gpr.GetScopedReg();
+  const auto reg_downcount = gpr.GetScopedReg();
+
+  LDR(IndexType::Unsigned, reg_downcount, PPC_REG, PPCSTATE_OFF(downcount));
+  MOVI2R(WA, 0);
+  CMP(reg_downcount, 0);                                          // if (downcount <= 0)
+  FixupBranch downcount_is_zero_or_negative = B(CCFlags::CC_LE);  // only do 1 invalidation; else:
+  LDR(IndexType::Unsigned, loop_counter, PPC_REG, PPCSTATE_OFF_SPR(SPR_CTR));
+  MOVI2R(reg_cycle_count, cycle_count_per_loop);
+  SDIV(WB, reg_downcount, reg_cycle_count);  // WB = downcount / cycle_count
+  SUB(WA, loop_counter, 1);                  // WA = CTR - 1
+  // ^ Note that this CTR-1 implicitly handles the CTR == 0 case correctly.
+  CMP(WB, WA);
+  CSEL(WA, WB, WA, CCFlags::CC_LO);  // WA = min(WB, WA)
+
+  // WA now holds the amount of loops to execute minus 1, which is the amount we need to adjust
+  // downcount, CTR, and Rb by to exit the loop construct with the right values in those
+  // registers.
+
+  // CTR -= WA
+  SUB(loop_counter, loop_counter, WA);
+  STR(IndexType::Unsigned, loop_counter, PPC_REG, PPCSTATE_OFF_SPR(SPR_CTR));
+
+  // downcount -= (WA * reg_cycle_count)
+  MSUB(reg_downcount, WA, reg_cycle_count, reg_downcount);
+  // ^ Note that this cannot overflow because it's limited by (downcount/cycle_count).
+  STR(IndexType::Unsigned, reg_downcount, PPC_REG, PPCSTATE_OFF(downcount));
+
+  SetJumpTarget(downcount_is_zero_or_negative);
+
+  if (IsDebuggingEnabled())
+  {
+    const ARM64Reg branch_watch = EncodeRegTo64(reg_cycle_count);
+    MOVP2R(branch_watch, &m_branch_watch);
+    LDRB(IndexType::Unsigned, WB, branch_watch, Core::BranchWatch::GetOffsetOfRecordingActive());
+    FixupBranch branch_over = CBZ(WB);
+
+    FixupBranch branch_in = B();
+    SwitchToFarCode();
+    SetJumpTarget(branch_in);
+
+    const BitSet32 gpr_caller_save =
+        gpr.GetCallerSavedUsed() &
+        ~BitSet32{DecodeReg(WB), DecodeReg(reg_cycle_count), DecodeReg(reg_downcount)};
```

Comment:
You should be able to add `loop_counter` to this bitset of registers that don't need saving. It previously needed saving because `ADD(loop_counter, WA, 1)` was emitted before this debugging code, but now it's emitted afterwards instead.

This also applies to Jit64.